server/api/src/rsExecMyRule.cpp

```diff
@@ -1,152 +1,143 @@
-                           _exec_inp->inpParamArray,
-                           out_param_desc);
-    if(inst_name.empty()) {
-        freeRErrorContent(&rei.rsComm->rError);
+    // Need to have a non zero inpParamArray for execMyRule to work.
+    if (!_exec_inp->inpParamArray) { // NOLINT(readability-implicit-bool-conversion)
+        // NOLINTNEXTLINE(cppcoreguidelines-owning-memory, cppcoreguidelines-no-malloc)
+        _exec_inp->inpParamArray = static_cast<MsParamArray*>(std::malloc(sizeof(MsParamArray)));
+        std::memset(_exec_inp->inpParamArray, 0, sizeof(MsParamArray));
     }
+    rei.msParamArray = _exec_inp->inpParamArray;
 
-    if(!inst_name.empty() && !err.ok()) {
-        rodsLog(
-            LOG_ERROR,
-            "%s : %d, %s",
-            __FUNCTION__,
-            err.code(),
-            err.result().c_str()
-        );
-        return err.code();
-    }
+    rstrcpy(rei.ruleName, EXEC_MY_RULE_KW, NAME_LEN); // NOLINT(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
 
-    trimMsParamArray( rei.msParamArray, _exec_inp->outParamDesc );
+    // NOLINTBEGIN(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+    const std::string my_rule_text = _exec_inp->myRule;
+    const std::string out_param_desc = _exec_inp->outParamDesc;
+    // NOLINTEND(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
 
-    *_out_arr = rei.msParamArray;
-    rei.msParamArray = NULL;
+    irods::rule_engine_context_manager<irods::unit, RuleExecInfo*, irods::AUDIT_RULE> re_ctx_mgr(
+        irods::re_plugin_globals->global_re_mgr, &rei);
+    irods::error err = re_ctx_mgr.exec_rule_text(inst_name, my_rule_text, _exec_inp->inpParamArray, out_param_desc);
 
-    if (!inst_name.empty() &&  err.code() < 0 ) {
-        rodsLog( LOG_ERROR,
-                 "rsExecMyRule : execMyRule error for %s, status = %d",
-                 _exec_inp->myRule, err.code() );
-        return err.code();
+    if (inst_name.empty()) {
+        freeRErrorContent(&rei.rsComm->rError);
     }
 
-    return err.code();
-}
+    if (!inst_name.empty() && !err.ok()) {
+        log_api::error("{}: {}, {}", __func__, err.code(), err.result());
+        return err.code(); // NOLINT(bugprone-narrowing-conversions, cppcoreguidelines-narrowing-conversions)
+    }
 
-int
-remoteExecMyRule( rsComm_t *_comm, execMyRuleInp_t *_exec_inp,
-                  msParamArray_t **_out_arr, rodsServerHost_t *rods_svr_host ) {
-    int status;
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+    trimMsParamArray(rei.msParamArray, _exec_inp->outParamDesc);
+
+    // TODO Copy the msparam array
+#if 0
+    *_out_param_arr = rei.msParamArray;
+#else
+    // NOLINTNEXTLINE(cppcoreguidelines-owning-memory, cppcoreguidelines-no-malloc)
+    *_out_param_arr = static_cast<MsParamArray*>(std::malloc(sizeof(MsParamArray)));
+    std::memset(*_out_param_arr, 0, sizeof(MsParamArray));
+    replMsParamArray(rei.msParamArray, *_out_param_arr);
+#endif
```

Comment:
`*_out_param_arr` points to the same `MsParamArray` as `_exec_inp->inpParamArray`. This is so weird, hence the else-block.

The else block gives the pointers different copies of the data. The input parameter is free'd, but the output pointer leaks memory.

Can't do anything about this until #6672.

Reply:
With #6672 being addressed by this PR, I've confirmed that the else-block is appropriate and results in no memory leaks.